sycl/doc/extensions/experimental/sycl_ext_oneapi_virtual_mem.asciidoc

```diff
@@ -0,0 +1,343 @@
+size_t get_recommended_mem_granularity(size_t numBytes, const physical_mem &syclPhysicalMem);
+
+} // namespace sycl::ext::oneapi::experimental
+```
+
+[frame="topbot",options="header,footer"]
+|=====================
+|Function |Description
+
+|`size_t get_minimum_mem_granularity(size_t numBytes, const device &syclDevice, const context &syclContext)` |
+Returns the minimum granularity of physical and virtual memory allocations of
+byte size `numBytes`.
+
+If `syclDevice` does not have `aspect::ext_oneapi_virtual_mem` the call throws
+an exception with `errc::feature_not_supported`.
+
+|`size_t get_minimum_mem_granularity(size_t numBytes, const queue &syclQueue)` |
+Same as `get_minimum_mem_granularity(numBytes, syclQueue.get_device(), syclQueue.get_context())`.
+
+|`size_t get_minimum_mem_granularity(size_t numBytes, const physical_mem &syclPhysicalMem)` |
+Same as `get_minimum_mem_granularity(numBytes, syclPhysicalMem.get_device(), syclPhysicalMem.get_context())`.
+
+|`size_t get_recommended_mem_granularity(size_t numBytes, const device &syclDevice, const context &syclContext)` |
+Returns the recommended granularity of physical and virtual memory allocations
+of byte size `numBytes`.
+
+If `syclDevice` does not have `aspect::ext_oneapi_virtual_mem` the call throws
+an exception with `errc::feature_not_supported`.
+
+|`size_t get_recommended_mem_granularity(size_t numBytes, const queue &syclQueue)` |
+Same as `get_recommended_mem_granularity(numBytes, syclQueue.get_device(), syclQueue.get_context())`.
+
+|`size_t get_recommended_mem_granularity(size_t numBytes, const physical_mem &syclPhysicalMem)` |
+Same as `get_recommended_mem_granularity(numBytes, syclPhysicalMem.get_device(), syclPhysicalMem.get_context())`.
+
+|=====================
+
+=== Reserving virtual address ranges
+
+Virtual address ranges are represented by a pointer and a number of bytes
+reserved for it. The pointer must be aligned in accordance with the minimum
+granularity, as queried through `get_minimum_mem_granularity`, and likewise the
+number of bytes must be a multiple of this granularity. It is the responsibility
+of the user to manage the constituents of any virtual address range they
+reserve.
+
+The interfaces for reserving, freeing, and manipulating the access mode of a
+virtual address range are defined as:
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+void *reserve_virtual_mem(const void *start, size_t numBytes, const context &syclContext);
+void *reserve_virtual_mem(size_t numBytes, const context &syclContext);
+
+void free_virtual_mem(const void* ptr, size_t numBytes, const context &syclContext);
+
+void set_access_mode(const void *ptr, size_t numBytes, access_mode mode, const context &syclContext);
+
+void set_inaccessible(const void *ptr, size_t numBytes, const context &syclContext);
+
+std::optional<access_mode> get_access_mode(const void *ptr, size_t numBytes, const context &syclContext);
+
+} // namespace sycl::ext::oneapi::experimental
+```
```

Comment:
I had a few thoughts about these APIs:

* I think maybe it would be better to use `uintptr_t` to represent a reserved address rather than `void *`.  There were two reasons.  First, it would help clarify that a reserved address may not be used as a pointer.  For example, you can't dereference it and you can't pass it to the explicit copy APIs.  Second, applications will probably need to do math on these addresses (e.g. adding an offset), and that's a lot more convenient if it's an integral type.

* I was thinking it might be better to define a new access mode enumeration, rather than reusing `sycl::access_mode`.  Again, there were two reasons.  It seems a little awkward to use `std::optional`.  If we defined a new enumeration, we could just have an enumerator named 'none`.  It also seems a bit awkward to document that only certain enumerator values from `sycl::access_mode` are allowed in these APIs.  Again, defining a new enumeration would solve this.

* I like the RAII design that you have for `physical_mem`, and I was wondering if you could do the same thing for reserved addresses.

These ideas are all independent, but if you put them together, you might have an API like this:

```
enum class address_access : /*unspecified*/ {
  none,
  read,
  read_write
};

class reserved_address {
 public:
  reserved_address(uintptr_t start, size_t numBytes, const context &syclContext,
                   address_access access = address_access::none);

  void set_access(address_access access);
  address_access get_access();
  uintptr_t get_start();
  size_t get_size();
  context get_context();
};
```

You would also make a change to the `physical_mem::map` function, returning a pointer:

```
class physical_mem {
 public:
  void *map(const reserved_address &reserved, size_t offset = 0);
  void *map(const reserved_address &reserved, address_access access, size_t offset = 0);
};
```

Returning a pointer from `map` make sense because you can use this pointer like a regular USM "device" pointer.  It can be dereferenced in a kernel and it can be passed to an explicit memory API from the host.

It wasn't clear to me if you are allowed to map a sub-range of a reserved address range, or if you need to map the whole range.  The API above makes sense if you need to map the whole range.  If it's OK to map a sub-range, then the API would look something like this:

```
class physical_mem {
 public:
  void *map(const reserved_address &reserved, size_t size, size_t reserved_offset, size_t physical_offset);
  void *map(const reserved_address &reserved, address_access access, size_t size, size_t reserved_offset, size_t physical_offset);
};
```

Reply:
I agree with both bullet 1 and 2. I will make those changes soon.

As for the `reserved_address` RAII class, I have some concerns. Currently the user must unmap address ranges before freeing them, so having the lifetime of these be tied to the RAII class may cause confusing errors if the class is released without an unmapping, potentially throwing in the destructor. An option would be to have either `physical_mem` track its mapped address ranges, having it unmap and release the remaining ones when it dies, or alternatively have `reserved_address` do the tracking, keeping any `physical_mem` it is mapped onto alive until it is destroyed itself. In both cases the user could untie them by calling the corresponding unmap. Note though that this does come with the overhead of keeping these relations.

Reply:
> alternatively have `reserved_address` do the tracking, keeping any `physical_mem` it is mapped onto alive until it is destroyed itself

When you map physical memory to a reserved address range, is it necessary to map the entire reserved address range?  If the answer is "yes", then the tracking overhead is very small because the `reserved_address` needs to track a maximum of 1 `physical_mem`.  To implement the tracking, you need to simply keep a copy of the `physical_mem` object inside of `reserved_address`.  The common reference semantics will do the rest.

If we go this route, I think I'd move the `map` function inside of `reserved_address`, which reinforces the idea that `reserved_address` holds a copy of the `physical_mem`.  It would also be convenient to add an `unmap` member function:

```
class reserved_address {
 public:
  /*...*/
  void *map(const pysical_mem &mem, size_t offset = 0);
  void unmap();
};
```

You raise an interesting point about errors that could occur during the destructor.  You have the same issue with `physical_mem` because the underlying backend calls might fail when deallocating the memory.  Luckily, both `physical_mem` and `reserved_address` are constructed with a `context`, which has an async error handler.  Therefore, we could specify that any errors during these destructor operations are reported as async errors.

Reply:
Sadly it is possible to map sub-ranges of a reserved address range onto different `physical_mem` objects. [`RecreateAddressRange` in sycl/test-e2e/VirtualMem/vector_with_virtual_mem.cpp](https://github.com/intel/llvm/pull/8954/files#diff-012963d223890adb166fa3d6ba3001a172fa571b2a7939336523dcad5dd13cc0R84-R113) shows an example of this, where the implementation fails to reserve the neighboring address range and instead creates an entirely new range and maps it onto all previously created `physical_mem` allocations.

Reply:
I'm wondering if `reserved_address` should still be responsible for tracking the `physical_mem` objects that are mapped to it, even though this could be a list.  There are two reasons:

* I think the code could have a short-circuit optimization for the case when there is only one `physical_mem` object, which I suspect will be a common case.
* If we did not do this, I wonder if the application would need to do the tracking anyway.  See my [comment here](https://github.com/intel/llvm/pull/8954/files#diff-012963d223890adb166fa3d6ba3001a172fa571b2a7939336523dcad5dd13cc0R84-R113) in your test application.  Therefore, adding the tacking to `reserved_address` wouldn't add any overhead. Instead, it would just shift overhead, and make the programming API easier.

Reply:
I tried implementing the `reserved_address` class as suggested, but I worry it would be too restrictive to make translating from low-level APIs to this extension difficult. A good example of this is the wide-range unmap mentioned in https://github.com/intel/llvm/pull/8954#discussion_r1164123315. Translating applications that do something like that would be difficult due to potential pointer arithmetics, and allowing it while also having `reserved_address` would make it difficult (and potentially costly) to correctly track the mappings, as discussed here.

Reply:
OK, makes sense.  If we agree that the first two bullet points above make sense, though, can we incorporate those changes?

Reply:
I have incorporated the first two bullets in the spec and implementation. Is that in line with what you had in mind?